### PR TITLE
Fix sundry bugs in the Veracruz virtual filesystem

### PR DESCRIPTION
**Pull request by gbryant-arm**
This fixes various bugs in the VFS that came to light in recent work on further Veracruz examples.

Bugs:
* wasm32 program fails when file descriptor returned by `path_open()` is bigger than 2**31
  * Fix: enforce upper bound of 2**31 to the result of `new_fd()`
* empty directory ("") conflicts with the root directory ("/") because forward slashes are stripped from the path prefix by `wasi-libc`
  * In WASI, directories are discovered (preopened) at startup, starting from file descriptor fd=3 and incrementing fd until until failure (cf. `__wasilibc_populate_preopens()` in wasi-libc).
An internal register indexes those file descriptors associated with their path.
This register is checked before calling `path_open()`, which requires the parent directory's file descriptor of the path to open.
Because any forward flash prefixing the path is stripped away when a directory is saved to the registry, the root directory ("/") ends up with the same prefix as the empty directory ("").
And since the most recent entries take precedence over the least recent ones, the result of determining the parent directory of a file under the root directory always defaults to the empty directory, which is not currently supported in our implementation of `path_open()`
  * Fix: remove empty directory from the VFS to avoid conflicts with the root directory

### CODE REVIEW

**Comment by veracruz-project-owner** [comment]
### AWS CodeBuild CI Report

* CodeBuild project: Veracruz
* Commit ID: e3230067fb483733904e9a0ffabb231b622ed6c6
* Result: FAILED
* [Build Logs](https://01evtei465.execute-api.eu-west-2.amazonaws.com/Prod/buildlogs?key=39bf61c9-688f-4f87-93d8-c1956958b49f%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*

File: execution-engine/src/fs.rs
Comment (dominic-mulligan-arm) [inline execution-engine/src/fs.rs:296]: I think this code can be improved a little, here, as you're calling `into()` on a `u32` value to produce an `Fd` and then immediately projecting out a `u32` value with the `new_fd.0` call below.  If you omit the `into()` here to produce only a `u32`, doesn't that simplify things?

**Review by dominic-mulligan-arm — COMMENTED** [review]
LGTM, though I'll wait for final approval from @ShaleXIONG when he is back before merging this, given the subject matter.

**Comment by veracruz-project-owner** [comment]
### AWS CodeBuild CI Report

* CodeBuild project: Veracruz
* Commit ID: 73992b9d137d936592891a5da787e8bc0b8bdcf3
* Result: SUCCEEDED
* [Build Logs](https://01evtei465.execute-api.eu-west-2.amazonaws.com/Prod/buildlogs?key=f17cd79b-8759-4b56-b32e-cfe55174614f%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*

**Comment by veracruz-project-owner** [comment]
### AWS CodeBuild CI Report

* CodeBuild project: Veracruz
* Commit ID: e8d2af188a390902c941c3a7b99dd2c3f42470e7
* Result: SUCCEEDED
* [Build Logs](https://01evtei465.execute-api.eu-west-2.amazonaws.com/Prod/buildlogs?key=4abc237b-dd89-4b28-b30c-dfaf92103870%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*

**Review by ShaleXIONG — APPROVED** [review]

**Pull request merged**
